pulsar/internal/buffer.go

```diff
@@ -217,3 +225,62 @@ func (b *buffer) Clear() {
 	b.readerIdx = 0
 	b.writerIdx = 0
 }
+
+type BufferPool interface {
+	// Get returns a cleared buffer if any is available, otherwise nil.
+	Get() *SharedBuffer
+
+	// Put puts the buffer back to the pool and available for other routines.
+	Put(*SharedBuffer)
+}
+
+type synchronizedBufferPool struct {
+	pool *sync.Pool
+}
+
+var _ BufferPool = synchronizedBufferPool{}
+
+func newBufferPool(pool *sync.Pool) synchronizedBufferPool {
+	return synchronizedBufferPool{pool: pool}
+}
+
+func (p synchronizedBufferPool) Get() *SharedBuffer {
+	buffer, ok := p.pool.Get().(Buffer)
+	if ok {
+		buffer.Clear()
+		return NewSharedBuffer(buffer)
+	}
+	return nil
+}
+
+func (p synchronizedBufferPool) Put(buffer *SharedBuffer) {
+	if !buffer.isCurrentlyShared.CompareAndSwap(true, false) {
+		// When swapping fails, it means that a previous go routine has already
+		// tried to recycle the buffer, indicating that the sharing is not there
+		// anymore and the buffer can safely be reused.
+		p.pool.Put(buffer)
+	}
```

Comment:
If:
- Buffer A created from pool
- Call A.Retain, create a new buffer B (isCurrentlyShared=true)
- Put back the A(isCurrentlyShared=false), the buffer A will be returned to the pool. And the buffer get recycled.
- Panic! The buffer B still holds this buffer.

Why not just use the ref count here?